File: _includes/code/java-v6/src/test/java/RBACTest.java
import io.weaviate.client6.v1.api.Authentication;
import io.weaviate.client6.v1.api.WeaviateClient;
import io.weaviate.client6.v1.api.rbac.AliasesPermission;
import io.weaviate.client6.v1.api.rbac.BackupsPermission;
import io.weaviate.client6.v1.api.rbac.ClusterPermission;
import io.weaviate.client6.v1.api.rbac.CollectionsPermission;
import io.weaviate.client6.v1.api.rbac.DataPermission;
import io.weaviate.client6.v1.api.rbac.GroupsPermission;
import io.weaviate.client6.v1.api.rbac.NodesPermission;
import io.weaviate.client6.v1.api.rbac.Permission;
import io.weaviate.client6.v1.api.rbac.ReplicatePermission;
import io.weaviate.client6.v1.api.rbac.Role;
import io.weaviate.client6.v1.api.rbac.RolesPermission;
import io.weaviate.client6.v1.api.rbac.TenantsPermission;
import io.weaviate.client6.v1.api.rbac.UsersPermission;
import io.weaviate.client6.v1.api.rbac.groups.GroupType;
import io.weaviate.client6.v1.api.rbac.users.DbUser;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

class RBACTest {

  private static WeaviateClient client;

  @BeforeAll
  public static void beforeAll() throws IOException {
    // START AdminClient
    // Connect to Weaviate as root user
    client = WeaviateClient.connectToLocal(config -> config
        // END AdminClient
        // Use custom port defined in tests/docker-compose-rbac.yml
        .port(8580)
        .grpcPort(50551)
        // START AdminClient
        .authentication(Authentication.apiKey("root-user-key")));
    // END AdminClient
    cleanup();
  }

  @AfterAll
  public static void afterAll() throws Exception {
    if (client != null) {
      client.close();
    }
  }

  public static void cleanup() throws IOException {
    // Clean up all test roles
    List<String> builtInRoles = List.of("admin", "root", "viewer", "read-only");
    var allRoles = client.roles.list();
    System.out.println(allRoles);
    for (Role role : allRoles) {
      if (!builtInRoles.contains(role.name())) {
        client.roles.delete(role.name());
      }
    }
    // Clean up all test users
    var allUsers = client.users.db.list();
    System.out.println("num of users:" + allUsers);
    for (DbUser user : allUsers) {
      if (!user.id().equals("root-user")) {
        client.users.db.delete(user.id());
      }
    }
  }

  @Test
  void testRolePermissionTypes() throws IOException {
    // START AddManageRolesPermission
    Permission[] rolesPermissions =
        new Permission[] {Permission.roles("testRole*", // Applies to all roles starting with "testRole"
            RolesPermission.Scope.MATCH, // Only allow role management with the current user's permission level
            // RolesPermission.Scope.ALL,  // Allow role management with all permissions
            RolesPermission.Action.CREATE, // Allow creating roles
            RolesPermission.Action.READ, // Allow reading roles
            RolesPermission.Action.UPDATE, // Allow updating roles
            RolesPermission.Action.DELETE // Allow deleting roles
        )};

    client.roles.create("testRole_ManageRoles", rolesPermissions);
    // END AddManageRolesPermission
    assertThat(client.roles.exists("testRole_ManageRoles")).isTrue();


    // START AddManageUsersPermission
    Permission[] usersPermissions =
        new Permission[] {Permission.users("testUser*", // Applies to all users starting with "testUser"
            UsersPermission.Action.CREATE, // Allow creating users
            UsersPermission.Action.READ, // Allow reading user info
            UsersPermission.Action.UPDATE, // Allow rotating user API key
            UsersPermission.Action.DELETE, // Allow deleting users
            UsersPermission.Action.ASSIGN_AND_REVOKE // Allow assigning and revoking roles to and from users
        )};

    client.roles.create("testRole_ManageUsers", usersPermissions);
    // END AddManageUsersPermission
    assertThat(client.roles.exists("testRole_ManageUsers")).isTrue();


    // START AddCollectionsPermission
    Permission[] collectionsPermissions =
        new Permission[] {Permission.collections("TargetCollection*", // Applies to all collections starting with "TargetCollection"
            CollectionsPermission.Action.CREATE, // Allow creating new collections
            CollectionsPermission.Action.READ, // Allow reading collection info/metadata
            CollectionsPermission.Action.UPDATE, // Allow updating collection configuration
            CollectionsPermission.Action.DELETE // Allow deleting collections
        ),};

    client.roles.create("testRole_ManageCollections", collectionsPermissions);
    // END AddCollectionsPermission
    assertThat(client.roles.exists("testRole_ManageCollections")).isTrue();


    // START AddTenantPermission
    Permission[] tenantsPermissions =
        new Permission[] {Permission.tenants("TargetCollection*", // Applies to all collections starting with "TargetCollection"
            "TargetTenant*", // Applies to all tenants starting with "TargetTenant"
            TenantsPermission.Action.CREATE, // Allow creating new tenants
            TenantsPermission.Action.READ, // Allow reading tenant info/metadata
            TenantsPermission.Action.UPDATE, // Allow updating tenant states
            TenantsPermission.Action.DELETE // Allow deleting tenants
        ),};

    client.roles.create("testRole_ManageTenants", tenantsPermissions);
    // END AddTenantPermission
    assertThat(client.roles.exists("testRole_ManageTenants")).isTrue();

    // TODO[g-despot] Tenants missing from Permission.data
    // START AddDataObjectPermission
    Permission[] dataPermissions =
        new Permission[] {Permission.data("TargetCollection*", // Applies to all collections starting with "TargetCollection"
            // "TargetTenant*", // Applies to all tenants starting with "TargetTenant"
            DataPermission.Action.CREATE, // Allow data inserts
            DataPermission.Action.READ, // Allow query and fetch operations
            DataPermission.Action.UPDATE, // Allow data updates
            DataPermission.Action.DELETE // Allow data deletes (set to false in example)
        ),};

    client.roles.create("testRole_ManageData", dataPermissions);
    // END AddDataObjectPermission
    assertThat(client.roles.exists("testRole_ManageData")).isTrue();


    // START AddBackupPermission
    Permission[] backupPermissions =
        new Permission[] {Permission.backups("TargetCollection*", // Applies to all collections starting with "TargetCollection"
            BackupsPermission.Action.MANAGE // Allow managing backups
        ),};

    client.roles.create("testRole_ManageBackups", backupPermissions);
    // END AddBackupPermission
    assertThat(client.roles.exists("testRole_ManageBackups")).isTrue();


    // START AddClusterPermission
    Permission[] clusterPermissions =
        new Permission[] {Permission.cluster(ClusterPermission.Action.READ), // Allow reading cluster data
        };

    client.roles.create("testRole_ReadCluster", clusterPermissions);
    // END AddClusterPermission
    assertThat(client.roles.exists("testRole_ReadCluster")).isTrue();


    // START AddNodesPermission
    Permission[] verbosePermissions =
        new Permission[] {Permission.nodes("TargetCollection*", // Applies to all collections starting with "TargetCollection"
            NodesPermission.Action.READ // Allow reading node metadata
        ),};

    // The `minimal` verbosity level is not exposed, use the standard one
    // which corresponds to 'verbose'
    client.roles.create("testRole_ReadNodes", verbosePermissions);
    // END AddNodesPermission
    assertThat(client.roles.exists("testRole_ReadNodes")).isTrue();


    // START AddAliasPermission
    Permission[] aliasPermissions =
        new Permission[] {Permission.aliases("TargetAlias*", // Applies to all aliases starting with "TargetAlias"
            "TargetCollection*", // Applies to all collections starting with "TargetCollection"
            AliasesPermission.Action.CREATE, // Allow alias creation
            AliasesPermission.Action.READ, // Allow listing aliases
            AliasesPermission.Action.UPDATE // Allow updating aliases
        // Delete is false in example
        ),};

    client.roles.create("testRole_ManageAliases", aliasPermissions);
    // END AddAliasPermission
    assertThat(client.roles.exists("testRole_ManageAliases")).isTrue();


    // START AddReplicationsPermission
    Permission[] replicatePermissions =
        new Permission[] {Permission.replicate("TargetCollection*", // Applies to all collections starting with "TargetCollection"
            "TargetShard*", // Applies to all shards starting with "TargetShard"
            ReplicatePermission.Action.CREATE, // Allow replica movement operations
            ReplicatePermission.Action.READ, // Allow retrieving replication status
            ReplicatePermission.Action.UPDATE // Allow cancelling replication operations
        // Delete is false in example
        ),};

    client.roles.create("testRole_ManageReplicas", replicatePermissions);
    // END AddReplicationsPermission
    assertThat(client.roles.exists("testRole_ManageReplicas")).isTrue();


    // START AddGroupsPermission
    Permission[] groupsPermissions =
        new Permission[] {Permission.groups("TargetGroup*", // Applies to all groups starting with "TargetGroup"
            GroupType.OIDC, GroupsPermission.Action.READ, // Allow reading group information
            GroupsPermission.Action.ASSIGN_AND_REVOKE // Allow assigning and revoking group memberships
        ),};

    client.roles.create("testRole_ManageGroups", groupsPermissions);
    // END AddGroupsPermission
    assertThat(client.roles.exists("testRole_ManageGroups")).isTrue();
  }

  @Test
  void testRoleLifecycle() throws IOException {
    String testRole = "testRole";
    String testUser = "custom-user";

    Permission[] initialPermissions = new Permission[] {Permission
        .collections("TargetCollection*", CollectionsPermission.Action.READ)};
    client.roles.create(testRole, initialPermissions);

    // START AddRoles
    Permission[] additionalPermissions = new Permission[] {
        Permission.data("TargetCollection*", DataPermission.Action.CREATE)};
    client.roles.addPermissions(testRole, additionalPermissions);
    // END AddRoles

    // START CheckRoleExists
    boolean exists = client.roles.exists(testRole);
    System.out.println(exists); // Returns True or False
    // END CheckRoleExists
    assertThat(exists).isTrue();

    // START InspectRole
    Role testRoleData = client.roles.get(testRole).orElse(null);
    System.out.println(testRoleData);
    System.out.println(testRoleData.permissions()); // Combined list
    // END InspectRole
    assertThat(testRoleData.permissions()).hasSize(2);
    assertThat(testRoleData.permissions()).contains(initialPermissions[0],
        additionalPermissions[0]);

    client.users.db.create(testUser);
    client.users.db.assignRoles(testUser, testRole);

    // START AssignedUsers
    var assignedUsers = client.roles.assignedUserIds(testRole);
    for (String user : assignedUsers) {
      System.out.println(user);
    }
    // END AssignedUsers
    assertThat(assignedUsers).contains(testUser);

    // START ListAllRoles
    List<Role> allRoles = client.roles.list();
    for (Role role : allRoles) {
      System.out.println(role.name() + " " + role);
    }
    // END ListAllRoles
    assertThat(allRoles).anyMatch(r -> r.name().equals(testRole));

    // START RemovePermissions
    Permission[] permissionsToRemove = new Permission[] {
        Permission.collections("TargetCollection*",
            CollectionsPermission.Action.READ),
        Permission.data("TargetCollection*", DataPermission.Action.CREATE),};
    client.roles.removePermissions(testRole, permissionsToRemove);
    // END RemovePermissions
    assertThat(client.roles.get(testRole).get().permissions()).isEmpty();

    // START DeleteRole
    client.roles.delete(testRole);
    // END DeleteRole
    assertThat(client.roles.exists(testRole)).isFalse();
  }

  @Test
  void testRoleExamples() throws IOException {
    String testUser = "custom-user";
    client.users.db.create(testUser);

    // START ReadWritePermissionDefinition
    // Define permissions (example confers read+write rights to collections starting with "TargetCollection")
    Permission[] rwPermissions = new Permission[] {
        // Collection level permissions
        Permission.collections("TargetCollection*",
            CollectionsPermission.Action.CREATE, // Allow creating new collections
            CollectionsPermission.Action.READ, // Allow reading collection info/metadata
            CollectionsPermission.Action.UPDATE, // Allow updating collection configuration
            CollectionsPermission.Action.DELETE // Allow deleting collections
        ),
        // Collection data level permissions
        Permission.data("TargetCollection*", DataPermission.Action.CREATE, // Allow data inserts
            DataPermission.Action.READ, // Allow query and fetch operations
            DataPermission.Action.UPDATE, // Allow data updates
            DataPermission.Action.DELETE // Allow data deletes (set to false in example)
        ),
        Permission.backups("TargetCollection*",
            BackupsPermission.Action.MANAGE),
        Permission.nodes("TargetCollection*", NodesPermission.Action.READ),
        Permission.cluster(ClusterPermission.Action.READ),};

    // Create a new role
    client.roles.create("rw_role", rwPermissions);
    // END ReadWritePermissionDefinition

    // START ReadWritePermissionAssignment
    // Assign the role to a user
    client.users.db.assignRoles(testUser, "rw_role");
    // END ReadWritePermissionAssignment

    var userRoles = client.users.db.assignedRoles(testUser);
    assertThat(userRoles).anyMatch(r -> r.name().equals("rw_role"));
    client.users.db.revokeRoles(testUser, "rw_role"); // Clean up for next assignment

    // START ViewerPermissionDefinition
    // Define permissions (example confers viewer rights to collections starting with "TargetCollection")
    Permission[] viewerPermissions = new Permission[] {
        Permission.collections("TargetCollection*",
            CollectionsPermission.Action.READ),
        Permission.data("TargetCollection*", DataPermission.Action.READ),};

    // Create a new role
    client.roles.create("viewer_role", viewerPermissions);
    // END ViewerPermissionDefinition

    // START ViewerPermissionAssignment
    // Assign the role to a user
    client.users.db.assignRoles(testUser, "viewer_role");
    // END ViewerPermissionAssignment
    userRoles = client.users.db.assignedRoles(testUser);
    assertThat(userRoles).anyMatch(r -> r.name().equals("viewer_role"));
    client.users.db.revokeRoles(testUser, "viewer_role"); // Clean up for next assignment

    // START MTPermissionsExample
    Permission[] mtPermissions =
        new Permission[] {Permission.tenants("TargetCollection*", // Applies to all collections starting with "TargetCollection"
            "TargetTenant*", // Applies to all tenants starting with "TargetTenant"
            TenantsPermission.Action.CREATE, // Allow creating new tenants
            TenantsPermission.Action.READ, // Allow reading tenant info/metadata
            TenantsPermission.Action.UPDATE, // Allow updating tenant states
            TenantsPermission.Action.DELETE // Allow deleting tenants
        ), Permission.data("TargetCollection*", // Applies to all collections starting with "TargetCollection"
            // "TargetTenant*", // Applies to all tenants starting with "TargetTenant"
            DataPermission.Action.CREATE, // Allow data inserts
            DataPermission.Action.READ, // Allow query and fetch operations
            DataPermission.Action.UPDATE, // Allow data updates
            DataPermission.Action.DELETE // Allow data deletes
        ),};

    // Create a new role
    client.roles.create("tenant_manager", mtPermissions);
    // END MTPermissionsExample

    // START MTPermissionsAssignment
    // Assign the role to a user
    client.users.db.assignRoles(testUser, "tenant_manager");
    // END MTPermissionsAssignment
    userRoles = client.users.db.assignedRoles(testUser);
    assertThat(userRoles).anyMatch(r -> r.name().equals("tenant_manager"));
  }

  @Test
  void testUserLifecycle() throws IOException {
    String testUser = "custom-user";
    String testRole = "testRole";

    client.users.db.delete(testUser);
    // START CreateUser
    String userApiKey = client.users.db.create(testUser);
    System.out.println(userApiKey);
    // END CreateUser
    assertThat(userApiKey).isNotNull().isNotEmpty();

    // START RotateApiKey
    String newApiKey = client.users.db.rotateKey(testUser);
    System.out.println(newApiKey);
    // END RotateApiKey
    assertThat(newApiKey).isNotNull().isNotEmpty().isNotEqualTo(userApiKey);

    Permission[] permissions = new Permission[] {Permission
        .collections("TargetCollection*", CollectionsPermission.Action.READ),};
    client.roles.create(testRole, permissions);

    // START AssignRole
    client.users.db.assignRoles(testUser, testRole, "viewer");
    // END AssignRole
    var roles = client.users.db.assignedRoles(testUser);
    assertThat(roles).extracting(Role::name).contains(testRole, "viewer");

    // START ListAllUsers
    var allUsers = client.users.db.list();
    System.out.println(allUsers);
    // END ListAllUsers
    assertThat(allUsers).anyMatch(u -> u.id().equals(testUser));

    // START ListUserRoles
    var userRoles = client.users.db.assignedRoles(testUser);
    for (Role role : userRoles) {
      System.out.println(role.name());
    }
    // END ListUserRoles
    assertThat(userRoles).extracting(Role::name).contains(testRole, "viewer");

    // START RevokeRoles
    client.users.db.revokeRoles(testUser, testRole);
    // END RevokeRoles
    roles = client.users.db.assignedRoles(testUser);
    assertThat(roles).extracting(Role::name)
        .doesNotContain(testRole)
        .contains("viewer");

    // START DeleteUser
    client.users.db.delete(testUser);
    // END DeleteUser
    assertThat(client.users.db.byName(testUser)).isEmpty();
  }
}

File: _includes/code/csharp/RBACTest.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Weaviate.Client.Models;
using Xunit;

namespace Weaviate.Client.Tests.RBAC;

public class RBACTest : IAsyncLifetime
{
    private WeaviateClient client;
    private const string RootUserKey = "root-user-key";

    public async Task InitializeAsync()
    {
        // START AdminClient
        // Connect to Weaviate as root user
        client = await Connect.Local(
            restPort: 8580,
            grpcPort: 50551,
            credentials: RootUserKey
        );
        // END AdminClient

        await Cleanup();
    }

    public Task DisposeAsync()
    {
        // C# client manages resources automatically, but we can run cleanup
        return Cleanup();
    }

    private async Task Cleanup()
    {
        // Clean up all test roles
        var builtInRoles = new List<string> { "admin", "root", "viewer", "read-only" };
        var allRoles = await client.Roles.ListAll();

        foreach (var role in allRoles)
        {
            if (!builtInRoles.Contains(role.Name))
            {
                await client.Roles.Delete(role.Name);
            }
        }

        // Clean up all test users
        var allUsers = await client.Users.Db.List();
        foreach (var user in allUsers)
        {
            if (user.UserId != "root-user")
            {
                await client.Users.Db.Delete(user.UserId);
            }
        }
    }

    [Fact]
    public async Task TestRolePermissionTypes()
    {
        // START AddManageRolesPermission
        var rolesPermissions = new PermissionScope[]
        {
            new Permissions.Roles("testRole*", RolesScope.Match) // Only allow role management with the current user's permission level
            {
                Create = true, // Allow creating roles
                Read = true,   // Allow reading roles
                Update = true, // Allow updating roles
                Delete = true  // Allow deleting roles
            }
        };

        await client.Roles.Create("testRole_ManageRoles", rolesPermissions);
        // END AddManageRolesPermission
        Assert.NotNull(await client.Roles.Get("testRole_ManageRoles"));


        // START AddManageUsersPermission
        var usersPermissions = new PermissionScope[]
        {
            new Permissions.Users("testUser*") // Applies to all users starting with "testUser"
            {
                Create = true, // Allow creating users
                Read = true,   // Allow reading user info
                Update = true, // Allow rotating user API key
                Delete = true, // Allow deleting users
                AssignAndRevoke = true // Allow assigning and revoking roles to and from users
            }
        };

        await client.Roles.Create("testRole_ManageUsers", usersPermissions);
        // END AddManageUsersPermission
        Assert.NotNull(await client.Roles.Get("testRole_ManageUsers"));


        // START AddCollectionsPermission
        var collectionsPermissions = new PermissionScope[]
        {
            new Permissions.Collections("TargetCollection*") // Applies to all collections starting with "TargetCollection"
            {
                Create = true, // Allow creating new collections
                Read = true,   // Allow reading collection info/metadata
                Update = true, // Allow updating collection configuration
                Delete = true  // Allow deleting collections
            }
        };

        await client.Roles.Create("testRole_ManageCollections", collectionsPermissions);
        // END AddCollectionsPermission
        Assert.NotNull(await client.Roles.Get("testRole_ManageCollections"));


        // START AddTenantPermission
        var tenantsPermissions = new PermissionScope[]
        {
            new Permissions.Tenants("TargetCollection*", "TargetTenant*") // Applies to specified collections/tenants
            {
                Create = true, // Allow creating new tenants
                Read = true,   // Allow reading tenant info/metadata
                Update = true, // Allow updating tenant states
                Delete = true  // Allow deleting tenants
            }
        };

        await client.Roles.Create("testRole_ManageTenants", tenantsPermissions);
        // END AddTenantPermission
        Assert.NotNull(await client.Roles.Get("testRole_ManageTenants"));


        // START AddDataObjectPermission
        var dataPermissions = new PermissionScope[]
        {
            new Permissions.Data("TargetCollection*", "TargetTenant*") // Applies to all collections starting with "TargetCollection"
            {
                Create = true, // Allow data inserts
                Read = true,   // Allow query and fetch operations
                Update = true, // Allow data updates
                Delete = true  // Allow data deletes
            }
        };

        await client.Roles.Create("testRole_ManageData", dataPermissions);
        // END AddDataObjectPermission
        Assert.NotNull(await client.Roles.Get("testRole_ManageData"));


        // START AddBackupPermission
        var backupPermissions = new PermissionScope[]
        {
            new Permissions.Backups("TargetCollection*") // Applies to all collections starting with "TargetCollection"
            {
                Manage = true // Allow managing backups
            }
        };

        await client.Roles.Create("testRole_ManageBackups", backupPermissions);
        // END AddBackupPermission
        Assert.NotNull(await client.Roles.Get("testRole_ManageBackups"));


        // START AddClusterPermission
        var clusterPermissions = new PermissionScope[]
        {
            new Permissions.Cluster { Read = true } // Allow reading cluster data
        };

        await client.Roles.Create("testRole_ReadCluster", clusterPermissions);
        // END AddClusterPermission
        Assert.NotNull(await client.Roles.Get("testRole_ReadCluster"));


        // START AddNodesPermission
        var verbosePermissions = new PermissionScope[]
        {
            new Permissions.Nodes("TargetCollection*", NodeVerbosity.Verbose) // Applies to all collections starting with "TargetCollection"
            {
                Read = true // Allow reading node metadata
            }
        };

        await client.Roles.Create("testRole_ReadNodes", verbosePermissions);
        // END AddNodesPermission
        Assert.NotNull(await client.Roles.Get("testRole_ReadNodes"));


        // START AddAliasPermission
        var aliasPermissions = new PermissionScope[]
        {
            new Permissions.Alias("TargetCollection*", "TargetAlias*")
            {
                Create = true, // Allow alias creation
                Read = true,   // Allow listing aliases
                Update = true  // Allow updating aliases
                // Delete is false by default
            }
        };

        await client.Roles.Create("testRole_ManageAliases", aliasPermissions);
        // END AddAliasPermission
        Assert.NotNull(await client.Roles.Get("testRole_ManageAliases"));


        // START AddReplicationsPermission
        var replicatePermissions = new PermissionScope[]
        {
            new Permissions.Replicate("TargetCollection*", "TargetShard*")
            {
                Create = true, // Allow replica movement operations
                Read = true,   // Allow retrieving replication status
                Update = true  // Allow cancelling replication operations
                // Delete is false by default
            }
        };

        await client.Roles.Create("testRole_ManageReplicas", replicatePermissions);
        // END AddReplicationsPermission
        Assert.NotNull(await client.Roles.Get("testRole_ManageReplicas"));


        // START AddGroupsPermission
        var groupsPermissions = new PermissionScope[]
        {
            new Permissions.Groups("TargetGroup*", RbacGroupType.Oidc)
            {
                Read = true, // Allow reading group information
                AssignAndRevoke = true // Allow assigning and revoking group memberships
            }
        };

        await client.Roles.Create("testRole_ManageGroups", groupsPermissions);
        // END AddGroupsPermission
        Assert.NotNull(await client.Roles.Get("testRole_ManageGroups"));
    }

    [Fact]
    public async Task TestRoleLifecycle()
    {
        string testRole = "testRole";
        string testUser = "custom-user";

        var initialPermissions = new PermissionScope[]
        {
            new Permissions.Collections("TargetCollection*") { Read = true }
        };
        await client.Roles.Create(testRole, initialPermissions);

        // START AddRoles
        var additionalPermissions = new PermissionScope[]
        {
            new Permissions.Data("TargetCollection*", "TargetTenant*") { Create = true }
        };
        await client.Roles.AddPermissions(testRole, additionalPermissions);
        // END AddRoles

        // START CheckRoleExists
        // In C#, we check by attempting to get the role
        var retrievedRole = await client.Roles.Get(testRole);
        bool exists = retrievedRole != null;
        Console.WriteLine(exists);
        // END CheckRoleExists
        Assert.True(exists);

        // START InspectRole
        var testRoleData = await client.Roles.Get(testRole);
        Console.WriteLine(testRoleData);
        // END InspectRole
        Assert.NotNull(testRoleData);
        Assert.Equal(2, testRoleData.Permissions.Count());

        // Check for presence of specific permission types
        Assert.Contains(testRoleData.Permissions, p => p is Permissions.Collections { Read: true });
        Assert.Contains(testRoleData.Permissions, p => p is Permissions.Data { Create: true });

        await client.Users.Db.Create(testUser);
        await client.Users.Db.AssignRoles(testUser, new[] { testRole });

        // START AssignedUsers
        var assignedUsers = await client.Roles.GetUserAssignments(testRole);
        foreach (var assignment in assignedUsers)
        {
            Console.WriteLine(assignment.UserId);
        }
        // END AssignedUsers
        Assert.Contains(assignedUsers, a => a.UserId == testUser);

        // START ListAllRoles
        var allRoles = await client.Roles.ListAll();
        foreach (var role in allRoles)
        {
            Console.WriteLine($"{role.Name} {role}");
        }
        // END ListAllRoles
        Assert.Contains(allRoles, r => r.Name == testRole);

        // START RemovePermissions
        var permissionsToRemove = new PermissionScope[]
        {
            new Permissions.Collections("TargetCollection*") { Read = true },
            new Permissions.Data("TargetCollection*", "TargetTenant*") { Create = true }
        };
        await client.Roles.RemovePermissions(testRole, permissionsToRemove);
        // END RemovePermissions

        var roleAfterRemove = await client.Roles.Get(testRole);
        Assert.Empty(roleAfterRemove.Permissions);

        // START DeleteRole
        await client.Roles.Delete(testRole);
        // END DeleteRole

        // Assert role is gone (Get throws NotFound or returns null depending on implementation, assuming similar to Exists check)
        // Based on provided Integration tests, Get throws NotFound when deleted if wrapped, or we check List
        await Assert.ThrowsAsync<WeaviateNotFoundException>(async () => await client.Roles.Get(testRole));
    }

    [Fact]
    public async Task TestRoleExamples()
    {
        string testUser = "custom-user";
        await client.Users.Db.Create(testUser);

        // START ReadWritePermissionDefinition
        // Define permissions (example confers read+write rights to collections starting with "TargetCollection")
        var rwPermissions = new PermissionScope[]
        {
            // Collection level permissions
            new Permissions.Collections("TargetCollection*")
            {
                Create = true, // Allow creating new collections
                Read = true,   // Allow reading collection info/metadata
                Update = true, // Allow updating collection configuration
                Delete = true  // Allow deleting collections
            },
            // Collection data level permissions
            new Permissions.Data("TargetCollection*", "TargetTenant*")
            {
                Create = true, // Allow data inserts
                Read = true,   // Allow query and fetch operations
                Update = true, // Allow data updates
                Delete = true  // Allow data deletes
            },
            new Permissions.Backups("TargetCollection*") { Manage = true },
            new Permissions.Nodes("TargetCollection*", NodeVerbosity.Verbose) { Read = true },
            new Permissions.Cluster { Read = true }
        };

        // Create a new role
        await client.Roles.Create("rw_role", rwPermissions);
        // END ReadWritePermissionDefinition

        // START ReadWritePermissionAssignment
        // Assign the role to a user
        await client.Users.Db.AssignRoles(testUser, new[] { "rw_role" });
        // END ReadWritePermissionAssignment

        var userRoles = await client.Users.Db.GetRoles(testUser);
        Assert.Contains(userRoles, r => r.Name == "rw_role");
        await client.Users.Db.RevokeRoles(testUser, new[] { "rw_role" });

        // START ViewerPermissionDefinition
        // Define permissions (example confers viewer rights to collections starting with "TargetCollection")
        var viewerPermissions = new PermissionScope[]
        {
            new Permissions.Collections("TargetCollection*") { Read = true },
            new Permissions.Data("TargetCollection*", "TargetTenant*") { Read = true }
        };

        // Create a new role
        await client.Roles.Create("viewer_role", viewerPermissions);
        // END ViewerPermissionDefinition

        // START ViewerPermissionAssignment
        // Assign the role to a user
        await client.Users.Db.AssignRoles(testUser, new[] { "viewer_role" });
        // END ViewerPermissionAssignment

        userRoles = await client.Users.Db.GetRoles(testUser);
        Assert.Contains(userRoles, r => r.Name == "viewer_role");
        await client.Users.Db.RevokeRoles(testUser, new[] { "viewer_role" });

        // START MTPermissionsExample
        var mtPermissions = new PermissionScope[]
        {
            new Permissions.Tenants("TargetCollection*", "TargetTenant*")
            {
                Create = true, // Allow creating new tenants
                Read = true,   // Allow reading tenant info/metadata
                Update = true, // Allow updating tenant states
                Delete = true  // Allow deleting tenants
            },
            new Permissions.Data("TargetCollection*", "TargetTenant*")
            {
                Create = true, // Allow data inserts
                Read = true,   // Allow query and fetch operations
                Update = true, // Allow data updates
                Delete = true  // Allow data deletes
            }
        };

        // Create a new role
        await client.Roles.Create("tenant_manager", mtPermissions);
        // END MTPermissionsExample

        // START MTPermissionsAssignment
        // Assign the role to a user
        await client.Users.Db.AssignRoles(testUser, new[] { "tenant_manager" });
        // END MTPermissionsAssignment

        userRoles = await client.Users.Db.GetRoles(testUser);
        Assert.Contains(userRoles, r => r.Name == "tenant_manager");
    }

    [Fact]
    public async Task TestUserLifecycle()
    {
        string testUser = "custom-user";
        string testRole = "testRole";

        try
        {
            await client.Users.Db.Delete(testUser);
        }
        catch { /* ignore if not exists */ }

        // START CreateUser
        string userApiKey = await client.Users.Db.Create(testUser);
        Console.WriteLine(userApiKey);
        // END CreateUser
        Assert.False(string.IsNullOrEmpty(userApiKey));

        // START RotateApiKey
        string newApiKey = await client.Users.Db.RotateApiKey(testUser);
        Console.WriteLine(newApiKey);
        // END RotateApiKey
        Assert.False(string.IsNullOrEmpty(newApiKey));
        Assert.NotEqual(userApiKey, newApiKey);

        var permissions = new PermissionScope[]
        {
            new Permissions.Collections("TargetCollection*") { Read = true }
        };
        await client.Roles.Create(testRole, permissions);

        // START AssignRole
        await client.Users.Db.AssignRoles(testUser, new[] { testRole, "viewer" });
        // END AssignRole

        var roles = await client.Users.Db.GetRoles(testUser);
        var roleNames = roles.Select(r => r.Name).ToList();
        Assert.Contains(testRole, roleNames);
        Assert.Contains("viewer", roleNames);

        // START ListAllUsers
        var allUsers = await client.Users.Db.List();
        Console.WriteLine(string.Join(", ", allUsers.Select(u => u.UserId)));
        // END ListAllUsers
        Assert.Contains(allUsers, u => u.UserId == testUser);

        // START ListUserRoles
        var userRoles = await client.Users.Db.GetRoles(testUser);
        foreach (var role in userRoles)
        {
            Console.WriteLine(role.Name);
        }
        // END ListUserRoles
        var userRoleNames = userRoles.Select(r => r.Name).ToList();
        Assert.Contains(testRole, userRoleNames);
        Assert.Contains("viewer", userRoleNames);

        // START RevokeRoles
        await client.Users.Db.RevokeRoles(testUser, new[] { testRole });
        // END RevokeRoles

        roles = await client.Users.Db.GetRoles(testUser);
        roleNames = roles.Select(r => r.Name).ToList();
        Assert.DoesNotContain(testRole, roleNames);
        Assert.Contains("viewer", roleNames);

        // START DeleteUser
        await client.Users.Db.Delete(testUser);
        // END DeleteUser

        var usersAfterDelete = await client.Users.Db.List();
        Assert.DoesNotContain(usersAfterDelete, u => u.UserId == testUser);
    }
}